### PR TITLE
bpo-38976: Add support for HTTP Only flag in MozillaCookieJar

### PR DESCRIPTION
**Pull request by jacobneiltaylor**
This PR adds support for the HttpOnly flag as encoded in CURL cookiejars.

This PR was mainly designed to allow the MozillaCookieJar to parse in the cookies, as previously they were considered comments and ignored. 

As HttpOnly is considered a non-standard attribute, the nonstandard attribute dict was considered the most appropriate place to persist this information.

https://bugs.python.org/issue38976

### CODE REVIEW

**Comment by jacobneiltaylor** [comment]
For more information:

- This is effectively the same issue discussed in https://bugs.python.org/issue2190
- While Mozilla bowsers no longer use this format, it is often used to store cookies for CLI tools
    - An example of this is cURL, which supports this functionality: https://github.com/curl/curl/blob/curl-7_39_0/lib/cookie.c#L644

File: Lib/http/cookiejar.py [hunk 1]
```diff
@@ -2004,6 +2004,7 @@ class MozillaCookieJar(FileCookieJar):
     header by default (Mozilla can cope with that).
 
     """
+    httponly_prefix = "#HttpOnly_"
```
Comment (orsenthil) [inline Lib/http/cookiejar.py]: We haven't visited this module often.  Do you think, we could move this class attributes to a  `__init__` method - The new httponly_prefix as well as the previous `magic_re` and `header`.

File: Lib/http/cookiejar.py [hunk 2]
```diff
@@ -2024,8 +2025,15 @@ def _really_load(self, f, filename, ignore_discard, ignore_expires):
         try:
             while 1:
                 line = f.readline()
+                rest = {}
+
                 if line == "": break
 
+                # detect httponly flag if present
```
Comment (orsenthil) [inline Lib/http/cookiejar.py]: Will a comment or a reference to an RFC help to explain why we are doing this?

**Review by orsenthil — COMMENTED** [review]
Thank you for your contribution, @jacobneiltaylor 

I only had two brief comments. Please see if you like to address them.

I also wonder _why_  a very old issue got resurfaced again? Is there a practical change in the curl and other clients that we missed noticing? (ref: https://bugs.python.org/msg300920)

File: Lib/http/cookiejar.py (hunk 1 above)
Comment (jacobneiltaylor) [reply Lib/http/cookiejar.py]: Just to clarify, you are asking for the class variables to be removed and the values moved into the constructor as literals?

Comment (orsenthil) [reply Lib/http/cookiejar.py]: Yes.

**Comment by jacobneiltaylor** [comment]
Thanks for the feedback @orsenthil - I'll work on those code changes.

With regard to the "why", I can share some limited details on the specific use case that prompted this contribution. An in-house SSO system I work with allows users to periodically vend libcurl-compatible cookiejar files. These files are explicitly for use with console scripts that interface with internal web APIs, usually for purposes of testing, experimentation and automation. They currently work with cURL-based BASH scripts, however there is increased interest in using them with more sophisticated Python-based automation.

The cookiejar files are generated with the `#HttpOnly_` prefix, which are ignored as comments under the current implementation. We have an internal workaround, but we thought it would be prudent to share a fix with the community in case others have the same or a similar use case (as niche as the use case may be).

File: Lib/http/cookiejar.py (hunk 1 above)
Comment (orsenthil) [reply Lib/http/cookiejar.py]: >  Do you think, we could move this class attributes to a ` __init__` method - The new httponly_prefix as well as the previous magic_re and header.

Hi Jacob, 

My original thinking was - we add a  `__init__` method and we move those variables to the `__init__` method, then our usage with `self.attr_name` becomes consistent with instance variable usages.  

Move them to module-level headers is fine too.
 
I notice that moving.compile` to the module header will add a slight performance to the module which was not present. I am fine with it, but I will try to measure the performance of the test suite, before and after to ensure there is no drastic difference with the change.

**Comment by orsenthil** [comment]
> Azure Pipelines PR — #20200402.24 failed 

this is `make patchcheck` failure, perhaps there is a extra-space or a newline some in the patch.  Running it locally could help us figure it out.

**Comment by jacobneiltaylor** [comment]
Thanks Senthil, that is now fixed by the looks of things.

With regard to the changes made, I opted to move all the class variables to module constants instead of literals. This was because they were either relatively complex declarations (multiline strings/compiled regex objects) or were referenced in multiple locations.

Additionally it is important to note that while HTTPOnly is a standard attribute in RFC6265, these modules are designed to model the cookies as defined in RFC2965.

As such, I still believe the nonstandard attributes dictionary is the correct location to persist this information.

File: Lib/http/cookiejar.py [hunk 3]
```diff
@@ -2004,28 +2012,29 @@ class MozillaCookieJar(FileCookieJar):
...
         try:
             while 1:
                 line = f.readline()
+                rest = {}
+
                 if line == "": break
 
+                # httponly is a cookie flag as defined in rfc6265
+                # when encoded in a netscape cookie file,
+                # the line is prepended with "#HttpOnly_"
```
Comment (orsenthil) [inline Lib/http/cookiejar.py:2033]: Verified this information here - https://curl.haxx.se/libcurl/c/CURLOPT_COOKIELIST.html

**Review by orsenthil — APPROVED** [review]

**Comment by dlenski** [comment]
This issue is a duplicate of bpo-2190, which means that my PR (https://github.com/python/cpython/pull/22798) for that one is probably now unnecessary.

File: Lib/http/cookiejar.py [hunk 4]
```diff
@@ -50,10 +50,18 @@ def _debug(*args):
         logger = logging.getLogger("http.cookiejar")
     return logger.debug(*args)
 
-
+HTTPONLY_ATTR = "HTTPOnly"
```
Comment (dlenski) [inline Lib/http/cookiejar.py:53]: This implementation is similar to my alternative approach in https://github.com/python/cpython/pull/22798 (and see bpo-2190), _except_:

- This version:
  - **Check if a cookie is "HTTP-only"** with `cookie.has_nonstandard_attr(http.cookiejar.HTTPONLY_ATTR))`
  - **Set a cookie as "HTTP-only"** with `cookie.set_nonstandard_attr("HTTPOnly", "")`
- My version: 
  - **Check** with `cookie.httponly`.
  - **Set** with `cookie.httponly = True`

**Pull request merged**